tests/integration/pilot/mcs/discoverability/discoverability_test.go

```diff
@@ -0,0 +1,238 @@
+	"istio.io/istio/pkg/test/scopes"
+	"istio.io/istio/pkg/test/util/retry"
+)
+
+const (
+	serviceA = "svc-a"
+	serviceB = "svc-b"
+)
+
+var (
+	i      istio.Instance
+	testNS string
+	echos  echo.Instances
+
+	retryTimeout = retry.Timeout(1 * time.Minute)
+)
+
+func TestMain(m *testing.M) {
+	framework.
+		NewSuite(m).
+		Label(label.CustomSetup).
+		RequireMinVersion(17).
+		RequireMinClusters(2).
+		Setup(func(ctx resource.Context) error {
+			crd, err := ioutil.ReadFile("../../testdata/mcs-serviceexport-crd.yaml")
+			if err != nil {
+				return err
+			}
+			return ctx.Config().ApplyYAML("", string(crd))
+		}).
+		Setup(istio.Setup(&i, func(ctx resource.Context, cfg *istio.Config) {
+			cfg.ControlPlaneValues = `
+values:
+  pilot:
+    env:
+      ENABLE_MCS_SERVICE_DISCOVERY: "true"`
+		})).
+		Setup(func(ctx resource.Context) error {
+			// Create a new namespace in each cluster.
+			ns, err := namespace.New(ctx, namespace.Config{
+				Prefix: "mcs",
+				Inject: true,
+			})
+			if err != nil {
+				return err
+			}
+			testNS = ns.Name()
+
+			// Create echo instances in each cluster.
+			echos, err = echoboot.NewBuilder(ctx).
+				WithClusters(ctx.Clusters()...).
+				WithConfig(echo.Config{
+					Service:   serviceA,
+					Namespace: ns,
+					Ports:     common.EchoPorts,
+				}).
+				WithConfig(echo.Config{
+					Service:   serviceB,
+					Namespace: ns,
+					Ports:     common.EchoPorts,
+				}).Build()
+			return err
+		}).
+		Run()
+}
+
+func TestClusterLocal(t *testing.T) {
+	framework.NewTest(t).
+		Features("traffic.mcs.servicediscovery").
+		Run(func(ctx framework.TestContext) {
+			dest := echos.GetOrFail(ctx, echo.Service(serviceB))
+			for index, c := range ctx.Clusters() {
+				src := echos.GetOrFail(ctx, echo.Service(serviceA).And(echo.InCluster(c)))
+
+				ctx.NewSubTestf("from cluster %d", index).RunParallel(func(ctx framework.TestContext) {
+					// Ensure all requests stay in the same cluster.
+					checkClustersReached(ctx, src, dest, cluster.Clusters{src.Config().Cluster})
+				})
+			}
+		})
+}
+
+func TestMeshWide(t *testing.T) {
+	framework.NewTest(t).
+		Features("traffic.mcs.servicediscovery").
+		Run(func(ctx framework.TestContext) {
+			// Export service B in all clusters.
+			createAndCleanupServiceExport(ctx, serviceB, ctx.Clusters())
+
+			dest := echos.GetOrFail(ctx, echo.Service(serviceB))
+			for index, c := range ctx.Clusters() {
+				src := echos.GetOrFail(ctx, echo.Service(serviceA).And(echo.InCluster(c)))
+
+				ctx.NewSubTestf("from cluster %d", index).RunParallel(func(ctx framework.TestContext) {
+					// Ensure that all clusters are reached.
+					checkClustersReached(ctx, src, dest, ctx.Clusters())
+				})
+			}
+		})
+}
+
+func TestServiceExportedInOneCluster(t *testing.T) {
+	framework.NewTest(t).
+		Features("traffic.mcs.servicediscovery").
+		Run(func(ctx framework.TestContext) {
+			// Only export service B in the first cluster.
+			cluster0 := ctx.Clusters()[0]
```

Comment:
use `ctx.Clusters().Kube()` to avoid issues related to real VM and other special cluster types.